src/contexts/regions/index.tsx

```diff
@@ -0,0 +1,143 @@
+import React, { createContext, useContext, useEffect, useState } from 'react';
+
+import {
+  decodeMaskString,
+  getBlockTimestamp,
+  parseHNString,
+} from '@/utils/functions';
+
+import {
+  HumanRegionId,
+  HumanRegionRecord,
+  RegionMetadata,
+  RegionOrigin,
+  RELAY_CHAIN_BLOCK_TIME,
+} from '@/models';
+
+import { useCoretimeApi, useRelayApi } from '../apis';
+import { ApiState } from '../apis/types';
+
+interface RegionsData {
+  regions: Array<RegionMetadata>;
+  config: {
+    timeslicePeriod: number;
+  };
+  loading: boolean;
+  updateRegionName: (_index: number, _name: string) => void;
+}
+
+const defaultRegionData: RegionsData = {
+  regions: [],
+  config: {
+    timeslicePeriod: 0,
+  },
+  loading: false,
+  updateRegionName: () => {
+    /** */
+  },
+};
+
+const RegionDataContext = createContext<RegionsData>(defaultRegionData);
+
+interface Props {
+  children: React.ReactNode;
+}
+
+const RegionDataProvider = ({ children }: Props) => {
+  const [regions, setRegions] = useState<Array<RegionMetadata>>([]);
+  const [timeslicePeriod, setTimeslicePeriod] = useState<number>(0);
+  const [loading, setLoading] = useState(false);
+
+  const {
+    state: { api: coretimeApi, apiState: coretimeApiState },
+  } = useCoretimeApi();
+  const {
+    state: { api: relayApi, apiState: relayApiState },
+  } = useRelayApi();
+
+  useEffect(() => {
+    if (
+      !coretimeApi ||
+      coretimeApiState !== ApiState.READY ||
+      !relayApi ||
+      relayApiState !== ApiState.READY
+    ) {
+      setRegions([]);
+      return;
+    }
+    const timeslicePeriod = parseHNString(
+      coretimeApi.consts.broker.timeslicePeriod.toString()
+    );
+    const fetchRegions = async (): Promise<void> => {
+      setLoading(true);
+
+      const _regions: Array<RegionMetadata> = [];
+      const res = await coretimeApi.query.broker.regions.entries();
+      for await (const [key, value] of res) {
+        const [regionId] = key.toHuman() as [HumanRegionId];
+        const regionData = value.toHuman() as HumanRegionRecord;
+
+        const { begin, core, mask } = regionId;
+        const { end, owner, paid } = regionData;
+
+        const beginBlockHeight = timeslicePeriod * parseHNString(begin);
+        const tsBegin = await getBlockTimestamp(relayApi, beginBlockHeight); // begin block timestamp
+
+        // rough estimation
+        const endBlockHeight = timeslicePeriod * parseHNString(end);
+        const tsEnd =
+          tsBegin +
+          (endBlockHeight - beginBlockHeight) * RELAY_CHAIN_BLOCK_TIME;
+
+        const nPaid = paid ? parseHNString(paid) : undefined;
+        const id = `${parseHNString(begin)}-${core}-${mask.slice(2)}`;
+        const name = localStorage.getItem(`region-${id}`);
+
+        _regions.push({
+          begin: tsBegin,
+          core,
+          mask: decodeMaskString(mask, Math.floor(timeslicePeriod / 8)),
```

Comment:
Do we need to decode the mask? Not sure if we are gonna use it for anything else other than for showing the core ownership property of each region. Even if we do I think it is best to show in in a hex format.

I think the simplest way to figure out the core ownership of a region would be to just divide `mask` / `FULL_MASK`.

Reply:
i don't think it's not a simple math.
we should count the number of bits set to 1.

Reply:
Why? We only want to know what proportion of the core can this region occupy.

Reply:
yup. the mask 0000ff00 and 00ff00 are essentially the same in terms of the number of available blocks.

Reply:
That is not what a core mask represents. Each polkadot core can process a fixed amount of work every 6 seconds. Currently each core is processing a single para block at a time. With Agile Coretime a single core can process blocks from many different parachains. The core mask determines how the parachains that share the same core share its resources. E.g. there might be two parachains assigned to a single where one of them utilize 40% and the other one utilizes 60% of the core’s capacity.

Reply:
You can think of all the parachains prior to Agile Coretime as having full core masks(each para has its own core)